webhook/webhook.go

```diff
@@ -265,8 +277,113 @@ func configureCerts(ctx context.Context, client kubernetes.Interface, options *C
+	logger := ac.Logger
+	key, quit := queue.Get()
+	if quit {
+		return endpointCheckShutdown
+	}
+	defer queue.Done(key)
+
+	item, exists, err := store.GetByKey(key.(string))
+	if err != nil || !exists {
+		return endpointCheckNotReady
+	}
+	endpoints, ok := item.(*corev1.Endpoints)
+	if !ok {
+		return endpointCheckNotReady
+	}
+	if len(endpoints.Subsets) == 0 {
+		logger.Warnf("%s/%v endpoint not ready: no subsets", namespace, name)
+		return endpointCheckNotReady
+	}
+	for _, subset := range endpoints.Subsets {
+		if len(subset.Addresses) > 0 {
+			return endpointCheckReady
+		}
+	}
+	logger.Warnf("%s/%v endpoint not ready: no ready addresses", namespace, name)
+	return endpointCheckNotReady
+}
+
+func CreateInformerEndpointSource(cl kubernetes.Interface, namespace, name string) cache.ListerWatcher {
+	return cache.NewListWatchFromClient(
+		cl.CoreV1().RESTClient(),
+		"endpoints",
+		namespace,
+		fields.ParseSelectorOrDie(fmt.Sprintf("metadata.name=%s", name)))
+}
+
+func (ac *AdmissionController) waitForEndpointReady(stopCh <-chan struct{}) (shutdown bool) {
+	logger := ac.Logger
+	defer func() {
+		if shutdown {
+			logger.Info("Endpoint readiness check stopped - controller shutting down")
+		} else {
+			logger.Infof("Endpoint %s/%s is ready", ac.Options.Namespace, ac.Options.ServiceName)
+		}
+	}()
+
+	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+	defer queue.ShutDown()
+	store, controller := cache.NewInformer(
+		CreateInformerEndpointSource(ac.Client, ac.Options.Namespace, ac.Options.ServiceName),
+		&corev1.Endpoints{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				if key, err := cache.MetaNamespaceKeyFunc(obj); err == nil {
+					queue.Add(key)
+				}
+			},
+			UpdateFunc: func(prev, curr interface{}) {
+				prevObj := prev.(*corev1.Endpoints)
+				currObj := curr.(*corev1.Endpoints)
+				if prevObj.ResourceVersion != currObj.ResourceVersion {
+					if key, err := cache.MetaNamespaceKeyFunc(curr); err == nil {
+						queue.Add(key)
+					}
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				if key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj); err == nil {
+					queue.Add(key)
+				}
+			},
+		},
+	)
+
+	controllerStopCh := make(chan struct{})
+	defer close(controllerStopCh)
+	go controller.Run(controllerStopCh)
+
+	if !cache.WaitForCacheSync(stopCh, controller.HasSynced) {
+		logger.Errorf("wait for cache sync failed")
+		return true
+	}
+
+	for {
+		select {
+		case <-stopCh:
+			return true
+		default:
+			ready := ac.endpointReady(store, queue, ac.Options.Namespace, ac.Options.ServiceName)
+			switch ready {
+			case endpointCheckShutdown:
+				return true
+			case endpointCheckReady:
+				return false
+			case endpointCheckNotReady:
+				// continue waiting for endpoint to be ready
+			}
+		}
+	}
+}
```

Comment:
I'm concerned about the amount of code we are using to solve this (without unit testing).

The initial implementation of this had a deadlock, and this basically uses an Endpoints informer to perform a readiness check, which itself has two problems:
1) Even Endpoints propagation is eventually consistent and no guarantee that the API Server will reach our pod.
2) Our webhooks don't currently define readiness probes, so they will report as ready instantly (before the http endpoint is available).

Let's discuss options more in the issue first?